br/pkg/streamhelper/advancer.go

```diff
@@ -214,121 +154,45 @@ func (c *CheckpointAdvancer) tryAdvance(ctx context.Context, rst RangesSharesTS)
 		return err
 	}
 
-	result, err := collector.Finish(ctx)
+	_, err = collector.Finish(ctx)
 	if err != nil {
 		return err
 	}
-	fr := result.FailureSubRanges
-	if len(fr) != 0 {
-		log.Debug("failure regions collected", zap.Int("size", len(fr)))
-		c.cache.InsertRanges(RangesSharesTS{
-			TS:     rst.TS,
-			Ranges: fr,
-		})
-	}
 	return nil
 }
 
+func tsoBefore(n time.Duration) uint64 {
+	now := time.Now()
+	return oracle.ComposeTS(now.UnixMilli()-n.Milliseconds(), 0)
+}
+
 // CalculateGlobalCheckpointLight tries to advance the global checkpoint by the cache.
 func (c *CheckpointAdvancer) CalculateGlobalCheckpointLight(ctx context.Context) (uint64, error) {
-	log.Info("[log backup advancer hint] advancer with cache: current tree", zap.Stringer("ct", c.cache))
-	rsts := c.cache.PopRangesWithGapGT(config.DefaultTryAdvanceThreshold)
-	if len(rsts) == 0 {
+	var targets []spans.Valued
+	c.checkpoints.TraverseValuesLessThan(tsoBefore(config.DefaultTryAdvanceThreshold), func(v spans.Valued) bool {
+		targets = append(targets, v)
+		return true
+	})
+	if len(targets) == 0 {
 		return 0, nil
 	}
-	samples := rsts
-	if len(rsts) > 3 {
-		samples = rsts[:3]
+	samples := targets
+	if len(targets) > 3 {
+		samples = targets[:3]
 	}
 	for _, sample := range samples {
-		log.Info("[log backup advancer hint] sample range.", zap.Object("range", sample.Zap()), zap.Int("total-len", len(rsts)))
+		log.Info("[log backup advancer hint] sample range.", zap.Stringer("sample", sample), zap.Int("total-len", len(targets)))
 	}
 
-	workers := utils.NewWorkerPool(uint(config.DefaultMaxConcurrencyAdvance), "regions")
-	eg, cx := errgroup.WithContext(ctx)
-	for _, rst := range rsts {
-		rst := rst
-		workers.ApplyOnErrorGroup(eg, func() (err error) {
-			return c.tryAdvance(cx, *rst)
-		})
-	}
-	err := eg.Wait()
+	err := c.tryAdvance(ctx, len(targets), func(i int) kv.KeyRange { return targets[i].Key })
 	if err != nil {
 		return 0, err
 	}
-	ts := c.cache.CheckpointTS()
+	ts := c.checkpoints.MinValue()
+	spans.Debug(c.checkpoints)
```

Comment:
will it be removed before pr is merged? :eyes:

Reply:
Well, that has been removed at the commit when subscription are supported, anyway I think we need to change it in this PR too...